server/repository/src/db_diesel/number_row.rs

```diff
@@ -55,6 +88,68 @@ impl<'a> NumberRowRepository<'a> {
+        store_id: &str,
+    ) -> Result<NextNumber, RepositoryError> {
+        // 1. First we try to just grab the next number from the database, in most cases this should work and be the fast.
+
+        // Note: Format string is used here because diesel does seems to support binding NumberRowType as a String or as it's own type.
+        // It's safe to use format here, as r#type is an enum with predefined values (No user input)
+        let update_query_str = format!(
+            r#"UPDATE number SET value = value+1 WHERE store_id = $1 and type = '{}' RETURNING value;"#,
+            r#type
+        );
+        let update_query = sql_query(update_query_str.clone()).bind::<Text, _>(store_id);
+
+        // Debug diesel query
+        // println!(
+        //     "{}",
+        //     diesel::debug_query::<crate::DBType, _>(&update_query).to_string()
+        // );
+        let update_result = update_query.get_result::<NextNumber>(&self.connection.connection);
+
+        match update_result {
+            Ok(result) => Ok(result),
+            Err(NotFound) => {
+                // 2. There was no record to update, so we need to insert a new one.
+
+                // We need to add an ON CONFLICT Clause for postgres just in case 2 threads insert at the same time (SQLite <on disk> does not need this as it only allows a single write transaction at a time).
+                // Without this postgres will throw a unique constraint violation error and rollback the transaction, which is hard to recover from, instead we just ignore the error and check if it returned a value
+                // It's safe to use format here, as these inputs are not user controlled
+                let insert_query_str = format!(
+                    r#"INSERT INTO number (id, value, store_id, type) VALUES ('{}', 1, $1, '{}') {} RETURNING value;"#,
+                    uuid(),
+                    r#type,
+                    ON_CONFLICT_DO_NOTHING
+                );
+
+                let insert_query = sql_query(insert_query_str).bind::<Text, _>(store_id);
+                let insert_result =
+                    insert_query.get_result::<NextNumber>(&self.connection.connection);
+
+                match insert_result {
+                    Ok(result) => Ok(result),
+                    Err(NotFound) => {
+                        // 3. If we got here another thread inserted the record before we we able to (we know this because nothing was returned for the insert)
+                        // We should now be able to do the same 'update returning' query as before to get our new number.
+
+                        let update_query =
+                            sql_query(update_query_str.clone()).bind::<Text, _>(store_id);
+
+                        match update_query.get_result::<NextNumber>(&self.connection.connection) {
+                            Ok(result) => Ok(result),
+                            Err(e) => Err(RepositoryError::from(e)),
```

Comment:
I think this match can be replace by just:

```rust
let result = update_query.get_result::<NextNumber>(&self.connection.connection)?;
Ok(result)
```

Reply:
Yes, that's better thanks!